embassy-stm32f4/src/rtc.rs

```diff
@@ -0,0 +1,286 @@
+
+            self.set_arr(arr);
+        });
+    }
+
+    unsafe fn on_interrupt(&mut self) {
+        // Clear interrupt flag
+        self.rtc.sr().write(|w| w.uif().clear_bit());
+
+        self.reset_timestamp();
+        self.recompute();
+        // self.rtc.deref().cr1.modify(|_, w| w.cen().set_bit());
+    }
+
+    fn trigger_alarm(&self, n: usize, cs: &CriticalSection) {
+        // self.rtc.intenclr.write(|w| unsafe { w.bits(compare_n(n)) });
+
+        let alarm = &self.alarms.borrow(cs)[n];
+        alarm.timestamp.set(u64::MAX);
+
+        // Call after clearing alarm, so the callback can set another alarm.
+        alarm.callback.get().map(|(f, ctx)| f(ctx));
+    }
+
+    fn set_alarm_callback(&self, n: usize, callback: fn(*mut ()), ctx: *mut ()) {
+        interrupt::free(|cs| {
+            let alarm = &self.alarms.borrow(cs)[n];
+            alarm.callback.set(Some((callback, ctx)));
+        })
+    }
+
+    fn set_alarm(&self, n: usize, alarm_timestamp: u64) {
+        self.rtc.cr1().modify(|_, w| w.cen().clear_bit());
+
+        interrupt::free(|cs| {
+            (&self.alarms.borrow(cs)[n]).timestamp.set(alarm_timestamp);
+        });
+
+        self.recompute();
+        self.rtc.cr1().modify(|_, w| w.cen().set_bit());
+    }
+
+    pub fn alarm0(&'static self) -> Alarm<T> {
+        Alarm { n: 0, rtc: self }
+    }
+    pub fn alarm1(&'static self) -> Alarm<T> {
+        Alarm { n: 1, rtc: self }
+    }
+    pub fn alarm2(&'static self) -> Alarm<T> {
+        Alarm { n: 2, rtc: self }
+    }
+}
+
+impl<T: Instance> embassy::time::Clock for RTC<T> {
+    fn now(&self) -> u64 {
+        self.timestamp + self.rtc.cnt().read().bits() as u64
+    }
+}
+
+pub struct Alarm<T: Instance> {
+    n: usize,
+    rtc: &'static RTC<T>,
+}
+
+impl<T: Instance> embassy::time::Alarm for Alarm<T> {
+    fn set_callback(&self, callback: fn(*mut ()), ctx: *mut ()) {
+        self.rtc.set_alarm_callback(self.n, callback, ctx);
+    }
+
+    fn set(&self, timestamp: u64) {
+        self.rtc.set_alarm(self.n, timestamp);
+    }
+
+    fn clear(&self) {
+        self.rtc.set_alarm(self.n, u64::MAX);
+    }
+}
+
+mod sealed {
+    pub trait Instance {}
+
+    impl Instance for crate::pac::TIM7 {}
+    impl Instance for crate::pac::TIM2 {}
+}
+
+/// Implemented by all RTC instances.
+pub trait Instance: sealed::Instance + Sized + 'static {
+    /// The interrupt associated with this RTC instance.
+    type Interrupt: OwnedInterrupt;
+
+    fn cr1(&self) -> tim6::CR1;
+    fn egr(&self) -> tim6::EGR;
+    fn psc(&self) -> tim6::PSC;
+    fn cnt(&self) -> tim6::CNT;
+    fn dier(&self) -> tim6::DIER;
+    fn arr(&self) -> tim6::ARR;
+    fn sr(&self) -> tim6::SR;
+
+    fn enable_clock();
+    fn ppre(clocks: Clocks) -> u8;
+    fn pclk(clocks: Clocks) -> u32;
+}
+
+impl Instance for crate::pac::TIM7 {
+    type Interrupt = interrupt::TIM7Interrupt;
+
+    fn arr(&self) -> tim6::ARR {
+        unsafe { (&(*crate::pac::TIM7::ptr())) }.arr
```

Comment:
cannot move out of a shared reference

move occurs because value has type `stm32f4::generic::Reg<u32, stm32f4xx_hal::stm32::tim6::_ARR>`, which does not implement the `Copy` traitrustc(E0507)